File: src/main/java/cl/ravenhill/keen/util/math/MinMax.java
package cl.ravenhill.keen.util.math;

import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;

import java.util.Comparator;
import java.util.function.Consumer;

import static java.lang.String.format;
import static java.util.Objects.requireNonNull;

public final class MinMax<C> implements Consumer<C> {

  private final Comparator<? super C> _comparator;

  private C _min;
  private C _max;
  private long _count = 0L;

  private MinMax(final Comparator<? super C> comparator) {
    _comparator = requireNonNull(comparator);
  }

  @Override
  public void accept(final C object) {
    _min = min(_comparator, _min, object);
    _max = max(_comparator, _max, object);
    ++_count;
  }

  @Contract("_ -> this")
  public MinMax<C> combine(@NotNull final MinMax<C> other) {
    _min = min(_comparator, _min, other._min);
    _max = max(_comparator, _max, other._max);
    _count += other._count;

    return this;
  }

  public long count() {
    return _count;
  }

  public C max() {
    return _max;
  }

  @Override
  public String toString() {
    return format("MinMax[count=%d, min=%s, max=%s]", _count, _min, _max);
  }

  public static <T> T min(final Comparator<? super T> comp, final T a, final T b) {
    return a != null ? b != null ? comp.compare(a, b) <= 0 ? a : b : a : b;
  }

  public static <T> T max(final Comparator<? super T> comp, final T a, final T b) {
    return a != null ? b != null ? comp.compare(a, b) >= 0 ? a : b : a : b;
  }

  @NotNull
  @Contract(value = "_ -> new", pure = true)
  public static <T> MinMax<T> of(final Comparator<? super T> comparator) {
    return new MinMax<>(comparator);
  }

  @NotNull
  @Contract(" -> new")
  public static <C extends Comparable<? super C>> MinMax<C> of() {
    return of(Comparator.naturalOrder());
  }
}

File: src/main/kotlin/cl/ravenhill/keen/util/math/MinMax.kt
package cl.ravenhill.keen.util.math

import org.jetbrains.annotations.Contract
import java.util.function.Consumer


class MinMax<C> private constructor(private val comparator: Comparator<in C>) : Consumer<C> {
    private var _min: C? = null
    private var _max: C? = null
    private var _count = 0L

    override fun accept(`object`: C) {
        _min = min(comparator, _min, `object`)
        _max = max(comparator, _max, `object`)
        ++_count
    }

    @Contract("_ -> this")
    fun combine(other: MinMax<C>): MinMax<C> {
        _min = min<C>(comparator, _min!!, other._min!!)
        _max = max<C>(comparator, _max!!, other._max!!)
        _count += other._count
        return this
    }

    fun count(): Long {
        return _count
    }

    fun max(): C? {
        return _max
    }

    override fun toString(): String {
        return String.format(
            "cl.ravenhill.keen.util.math.MinMax[count=%d, min=%s, max=%s]",
            _count,
            _min,
            _max
        )
    }

    companion object {
        fun <T> min(comp: Comparator<in T>, a: T?, b: T?): T? {
            return if (a != null) if (b != null) if (comp.compare(a, b) <= 0) a else b else a else b
        }

        fun <T> max(comp: Comparator<in T>, a: T?, b: T?): T? {
            return if (a != null) if (b != null) if (comp.compare(a, b) >= 0) a else b else a else b
        }

        @Contract(value = "_ -> new", pure = true)
        fun <T> of(comparator: Comparator<in T>): MinMax<T> {
            return MinMax(comparator)
        }

        @JvmStatic
        @Contract(" -> new")
        fun <C : Comparable<C>?> of(): MinMax<C> {
            return of(Comparator.naturalOrder())
        }
    }
}